x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4HttpServerTransportCloseNotifyTests.java

```diff
@@ -0,0 +1,253 @@
+import io.netty.handler.codec.http.HttpRequestEncoder;
+import io.netty.handler.codec.http.HttpResponseDecoder;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.util.concurrent.Future;
+
+import org.elasticsearch.common.network.NetworkService;
+import org.elasticsearch.common.settings.MockSecureSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.env.TestEnvironment;
+import org.elasticsearch.http.AbstractHttpServerTransportTestCase;
+import org.elasticsearch.http.HttpServerTransport;
+import org.elasticsearch.http.netty4.Netty4HttpServerTransport;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.telemetry.tracing.Tracer;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.netty4.SharedGroupFactory;
+import org.elasticsearch.transport.netty4.TLSConfig;
+import org.elasticsearch.xpack.core.ssl.SSLService;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.net.ssl.SSLException;
+
+public class SecurityNetty4HttpServerTransportCloseNotifyTests extends AbstractHttpServerTransportTestCase {
+
+    private static void shutdownAll(HttpServer server, HttpClient client) {
+        server.netty.stop();
+        server.threadPool.shutdownNow();
+        safeAwait(client.netty.config().group().shutdownGracefully());
+    }
+
+    private static <T> T safePoll(BlockingQueue<T> queue) {
+        try {
+            var t = queue.poll(5, TimeUnit.SECONDS);
+            if (t == null) {
+                throw new AssertionError("queue is empty");
+            } else {
+                return t;
+            }
+        } catch (Exception e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    private static <T> T safeAwait(Future<T> nettyFuture) {
+        try {
+            return nettyFuture.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    private HttpServer setupHttpServer(String tlsProtocols) {
+        var cert = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.crt");
+        var pk = getDataPath("/org/elasticsearch/xpack/security/transport/ssl/certs/simple/testnode.pem");
```

Comment:
Aha nice find! Although it seems hardly ideal to be using a cert which exposed in 2019...

Reply:
I tried to use `ssc = new SelfSignedCertificate();` from netty, it relies on [bouncy-castle](https://mvnrepository.com/artifact/org.bouncycastle/bcpkix-jdk18on), but we dont have it in dependencies. Should I update deps or static cert?

Reply:
I think we have good reasons not to rely on bouncy castle in prod code but it seems like it should be ok for tests IMO. If it's easy to add then please do so and we can raise it for discussion as the PR nears completion.

Reply:
Made it work with few security suppressions:
```
@ESTestCase.WithoutSecurityManager
@SuppressForbidden(reason = "requires java.io.File for netty self-signed certificate")
```